### ISSUE TITLE
Reading virtual references back out into VirtualiZarr Manifests

### ISSUE THREAD

**Issue by TomNicholas** [earth-mover/icechunk#104]
I would like to be able to read virtual references back out from an icechunk store into VirtualiZarr `ManifestArray` objects.

Note this issue is the icechunk equivalent of https://github.com/zarr-developers/VirtualiZarr/issues/118, which is about reading kerchunk references into `ManifestArray` objects.

The main use case is appending new data to an existing store (see https://github.com/zarr-developers/VirtualiZarr/issues/21#issuecomment-2114001053), so that when some new data arrives (e.g. a new grib file with today's weather data), I can add an updated snapshot just with something like:

```python
import virtualizarr as vz

# avoids re-extracting all the metadata from all the past grib files, so should be quick
existing_vds = vz.open_virtual_dataset(icechunkstore, reader='icechunk')

new_vds = vz.open_virtual_dataset('todays_weather.grib', reader='grib')

updated_vds = xr.concat([existing_vds, new_vds], dim='time')

# commit new snapshot that includes today's data
# requires https://github.com/earth-mover/icechunk/issues/103
updated_vds.virtualize.to_icechunk(icechunkstore)
icechunkstore.commit('<todays-date>')
```

In order to implement that Icechunk reader for virtualizarr I would need some API for getting all virtual (and non-virtual) references for a snapshot back out of the Icechunk store, ideally as a `vz.ManifestArray` or something I can cheaply coerce to one (see [`ChunkManifest.from_arrays()`](https://github.com/zarr-developers/VirtualiZarr/blob/47a5e8702e44f71bb355bcba0ff6214fe6d09d83/virtualizarr/manifests/manifest.py#L142)).

Writing the updated references as a new snapshot also requires https://github.com/earth-mover/icechunk/issues/103.

(I guess the `.virtualize.to_icechunk` method might also need to know to do `array.resize` in this example... (see the Append example in [this notebook](https://github.com/earth-mover/icechunk/blob/main/icechunk-python/notebooks/demo-dummy-data.ipynb).)

Running that above snippet as a cron job / event-driven serverless function should go a long way towards making ingestion of regularly-updated data archives easier. (cc @mpiannucci)

This feature might also be useful to allow using icechunk as a serialization format during large serverless reductions (xref https://github.com/zarr-developers/VirtualiZarr/issues/123).

cc @paraseba

**Comment by abarciauskas-bgse** [comment]
Should we add a note that it is not currently possible to append virtual refs to icechunk stores in https://icechunk.io/icechunk-python/virtual/ and point to this issue? cc @mpiannucci

**Comment by abarciauskas-bgse** [comment]
Ok thanks for explaining how the append method is working @TomNicholas - but if I understand correctly there is no way to append virtual refs at the moment, so I would still see a benefit to others to make this clear in the documentation.

Is there any plans to work on this feature in VirtualiZarr @TomNicholas @mpiannucci ?

**Comment by TomNicholas** [comment]
> I would still see a benefit to others to make this clear in the documentation.

:+1:

> Is there any plans to work on this feature in `VirtualiZarr`

I would very much like to see it, but don't personally need it (yet) so will be prioritizing other things. It's a good thing for somebody else to add, especially someone who is (or wants to become) familiar with xarray's existing `to_zarr` implementation.

**Comment by TomNicholas** [comment]
I'm going to use https://github.com/zarr-developers/VirtualiZarr/issues/21 to track the `append_dim` idea in VirtualiZarr.

Separately I still think that being able to pull virtual references back out would be useful. Imagine for example someone makes a public icechunk store containing virtual references, then someone from a different organisation wants to refer to those references - they would need to pull them out. (Whether or not people creating catalogs like this is a good idea is another question, but I bet people will ask for that.)

**Comment by TomNicholas** [comment]
@rabernat just said in the Pangeo showcase that icechunk is not yet at a state where it guarantees that references written with the current spec will be readable even with future versions of this library. To me that makes the feature of reading virtual references back out into VirtualiZarr actually quite important to have sooner rather than later. If that existed, then if the icechunk store changed on disk, I could always roll back to a version of the icechunk library that could read the references in order to suck them out then re-save them in the new format.

**Comment by mpiannucci** [comment]
Cna you propose what the ideal API for this would be? It gets confusing for me thinking about how there can be mixed virtual and native chunks

**Comment by TomNicholas** [comment]
> It gets confusing for me thinking about how there can be mixed virtual and native chunks

Yeah good point. All variables in an icechunk store are potentially available as actual loaded data, or as references, so I would say that there should be a new virtualizarr reader, such that 

```python
vds = open_virtual_dataset(icechunkstore, loadable_variables=...)
```
returns an `xr.Dataset` containing (lazy) numpy arrays for the `loadable_variables`, and `ManifestArrays` for all the other variables.

I thought maybe we could imagine this function returning virtual vs native variables based on how they were stored in icechunk, but IIUC that's actually not possible right now because a single zarr array could be backed by some virtual and some native chunks, which VirtualiZarr has no way to represent in one array.

The API needed for icechunk-python at its most basic is just `.get_virtual_ref`, the inverse of `.set_virtual_ref`. Again it would be more efficient to get back a whole `ManifestArray` of chunk references in one call but that's just a performance optimization.

**Comment by mpiannucci** [comment]
Cool. Would it return None if the chunk is native? Or throw an error?

**Comment by paraseba** [comment]
I wanted to mention that having mixed virtual and non virtual chunks in a single array, is _exactly_ what we want to enable: people starting from a virtual dataset and, with time, update the dataset in Icechunk (which would generate non-vrtual chunks). So we expect this case to be very common.

**Comment by paraseba** [comment]
I support adding `get_virtual_ref`, but it could be more useful to have `get_chunk_ref` that gives you the whole information, including the fact that it could be Virtual, Inline or "Normal". Then the client code can filter as needed. For reference, this is what `get_chunk_ref` returns in Rust:
```rust
pub enum ChunkPayload {
    Inline(Bytes),
    Virtual(VirtualChunkRef),
    Ref(ChunkRef),
}
```

**Comment by TomNicholas** [comment]
> Would it return `None` if the chunk is native?

No, it would still return the virtual reference to that chunk, the reference would just refer to deep within this icechunk store. IIUC icechunk chunks are immutable once written, so referring to a chunk in an icechunk store from elsewhere is exactly as reliable as referring to a chunk in a traditional zarr store / legacy-formatted file.

> I support adding `get_virtual_ref` but it could be more useful to have `get_chunk_ref` that gives you the whole information, including the fact that it could be Virtual, Inline or "Normal".

That would be fine for `VirtualiZarr` too (and maybe more useful for reasons I can't think of yet).

**Comment by paraseba** [comment]
Inline chunks I think are the biggest issue here. As @TomNicholas says, "normal" chunks can also be seen as virtual chunks, pointing to the icechunk store. But inline ones live only in the manifest. We _could_ provide a way to reference them via an offset/length, but it would be pretty tricky and ugly.

**Comment by paraseba** [comment]
> Tbh this makes inlined chunks seem like a leaky abstraction.
They are, I don't love them, they are leaky and probably a premature optimization. 

We could make IC not use inline chunks unless the user overrides a config, I suppose.

**Comment by mpiannucci** [comment]
Inline chunks don't matter to virtualizarr at all. To virtualizarr it's the same as a native chunk

**Comment by TomNicholas** [comment]
Except we're talking about here having native chunks also be extracted as virtual references, which @paraseba is saying is harder to provide for inclined chunks than for native chunks.

**Comment by TomNicholas** [comment]
Except we're talking about here having native chunks also be extracted as virtual references, which @paraseba is saying is harder to provide for inclined chunks than for native chunks.

> They are, I don't love them, they are leaky and probably a premature optimization.

It's unclear to me that they are necessary at all. Why can't you just "inline" a variable by storing it as a single native chunk? That requires an extra get request I guess? Or because then you can't cache it in the rust client?

**Comment by paraseba** [comment]
> That requires an extra get request I guess?

This. They are useful for things like coordinate arrays.

**Comment by mpiannucci** [comment]
> Except we're talking about here having native chunks also be extracted as virtual references, which @paraseba is saying is harder to provide for inclined chunks than for native chunks.

Got itttt sorry I misunderstood :)

**Comment by maxrjones** [comment]
Hey folks, thanks for all your work and considering these issues! I just wanted to share that I view the primary use case for this feature to be debugging. E.g., It would be nice to be able to investigate https://github.com/developmentseed/mursst-icechunk-updater/issues/23 by re-parsing the appended data into a ManifestStore and looking at the paths, byte ranges, and checksums relative to the expected values.

**Comment by TomNicholas** [comment]
@mpiannucci started work on the API that IC would need over in #1307, but I think this needs more thought, for a few reasons.

1. **What happens if only some of the references in an array are virtual?** The current implementation of `all_virtual_refs` will return only the virtual references, effectively leaving gaps in the array. If we naively built a `virtualizarr.ManifestArray` from those refs, virtualizarr would treat the absent native chunks as missing. Instead we need Icechunk to return the location and byte range of the native chunks alongside the virtual ones. It might be useful to be able to tell them apart on the virtualizarr side, but we need all of them. They are valid things for virtual references to point at, since they are immutable (until garbage collected).

2. **Returning virtual references on a per-array basis would be more convenient.** A more useful interface for virtualizarr would be if icechunk provided something like `.list_arr_virtual_refs` (or maybe `.list_virtual_refs_in_prefix`), because each call to that method would translate to one `virtualizarr.ManifestArray`. This is minor though - parsing the keys returned by a single `all_virtual_refs` call would also work, as would using zarr's `list_prefix` then calling `.get_virtual_ref` many times.

3. **How can we handle inlined chunks?** (This is discussed above.) In theory any chunk in Icechunk could also be inlined into a manifest. VirtualiZarr still needs to somehow know about these chunks too. Ideally virtualizarr would also be given the location and byte range of these parts of the manifest file, but is that even possible with the compressed flatbuffers files? A fallback might be for virtualizarr to directly load arrays with any inlined chunks, but that requires a way for the icechunk client to know which arrays contain at least one inlined reference, which is leaking through the storage abstraction.

**Comment by jbusecke** [comment]
Just ran into this issue yesterday with a concrete usecase I wanted to share: 

Over at https://github.com/developmentseed/ismip-indexing/pull/1 @abarciauskas-bgse has been working on virtualizing part of the ISMIP archive (related to CMIP6, but the details are both a bit complicated and also not really relevant here). 

Our goal is to present the data users with two different zarr structures:
- One where each group contains a single variable (equivalent to a single variable dataset in xarray) for a given model experiement (this is conform with the way CMIP data is delivered)
- Another non-conform but likely more performant and user friendly version where we combine all the variables of a given model and experiment into a single group. My hope in particular is that this will make it feasible to open the entire store with datatree by reducing the severity of some of the issues described in https://github.com/pydata/xarray/issues/10579 and https://github.com/pydata/xarray/issues/9511

However the initial virtualization takes quite long, and my hope in the future would be that I could simply recreate a ManifestArray backed datatree, modify the structure, and rewrite it to a second icechunk store. I personally think that this would open all kinds of cool use cases where any particular user group can take an existing virtual store (fast) and apply e.g. metadata modifications, reorganize or filter the data to their needs without having to go through an expensive compute step again.

**Comment by TomNicholas** [comment]
I think if we had #1549 and #1469 then we could do this fairly easily.

EDIT: Actually it's still hard: https://github.com/earth-mover/icechunk/issues/1549#issuecomment-3730488536

**Comment by TomNicholas** [comment]
@paraseba here's one suggestion for how to make this work. We add this new API to IC:

Rust — one method on `icechunk::Store`:

```rust
impl Store {
    pub async fn array_chunk_refs(
        &self,
        array_path: &Path,
        native_chunks_prefix: &str,  // the location in storage of the native chunks directory
    ) -> StoreResult<ArrayChunkRefs>;
}

pub struct ArrayChunkRefs {
    pub paths:   ndarray::Array<String, IxDyn>,  // shape = chunk grid; "" = missing, "__inlined__" = inline
    pub offsets: ndarray::Array<u64, IxDyn>,
    pub lengths: ndarray::Array<u64, IxDyn>,
    pub kind:    ndarray::Array<u8, IxDyn>,      // 0=missing, 1=virtual, 2=native, 3=inline
    pub inlined: HashMap<Vec<u32>, Bytes>,       // sparse, only inline slots
}
```

Internally iterates `Session::all_node_chunks` for the array; expands `vcc://` virtual locations via the session's resolver; renders native chunks as `format!("{native_chunks_prefix}/{chunk_id}")`.

Python — same on `icechunk.IcechunkStore`:

```
class IcechunkStore:
    async def array_chunk_refs(
        self,
        array_path: str,
        native_chunks_prefix: str,
    ) -> ArrayChunkRefs: ...

class ArrayChunkRefs:
    paths:   np.ndarray  # StringDType, shape=grid_shape
    offsets: np.ndarray  # uint64
    lengths: np.ndarray  # uint64
    kind:    np.ndarray  # uint8
    inlined: dict[tuple[int, ...], bytes]
```

I think this is the minimum that VZ needs to be able to translate to `ManifestArrays`. This handles arbitrary mixtures of inline/native/virtual chunks.

It's basically the inverse of `.set_virtual_refs_arr` - it's also quite leaky and gross, but gives VZ everything it needs in one method, with no new dependencies.

**Comment by TomNicholas** [comment]
This was closed by https://github.com/zarr-developers/VirtualiZarr/pull/991 and released in VirtualiZarr v2.6.2

(also cc @vladidobro who I remember wanted this to be able to manipulate `ManifestArrays` in-memory)
